zkevm-circuits/src/root_circuit.rs

```diff
@@ -0,0 +1,223 @@
+
+#[cfg(any(feature = "test", test))]
+pub use aggregation::TestAggregationCircuit;
+
+/// RootCircuit for aggregating SuperCircuit into a much smaller proof.
+pub struct RootCircuit<'a, M: MultiMillerLoop> {
+    svk: KzgSvk<M>,
+    snark: SnarkWitness<'a, M::G1Affine>,
+    instances: Vec<M::Scalar>,
+}
+
+impl<'a, M: MultiMillerLoop> RootCircuit<'a, M>
+where
+    M::G1Affine: SerdeObject,
+    M::G2Affine: SerdeObject,
+{
+    /// Create a `RootCircuit` with accumulator computed given a `SuperCircuit`
+    /// proof and its instances. Returns `None` if given proof is invalid.
+    pub fn new(
+        params: &ParamsKZG<M>,
+        super_circuit_protocol: &'a PlonkProtocol<M::G1Affine>,
+        super_circuit_instances: Value<&'a Vec<Vec<M::Scalar>>>,
+        super_circuit_proof: Value<&'a [u8]>,
+    ) -> Result<Self, snark_verifier::Error> {
+        let num_instances = super_circuit_protocol.num_instance.iter().sum::<usize>() + 4 * LIMBS;
+        let instances = {
+            let mut instances = Ok(vec![M::Scalar::zero(); num_instances]);
+            super_circuit_instances
+                .as_ref()
+                .zip(super_circuit_proof.as_ref())
+                .map(|(super_circuit_instances, super_circuit_proof)| {
+                    let snark = Snark::new(
+                        super_circuit_protocol,
+                        super_circuit_instances,
+                        super_circuit_proof,
+                    );
+                    instances = aggregate::<M>(params, [snark]).map(|accumulator_limbs| {
+                        iter::empty()
+                            // Propagate `SuperCircuit`'s instances
+                            .chain(super_circuit_instances.iter().flatten().cloned())
+                            // Output aggregated accumulator limbs
+                            .chain(accumulator_limbs)
+                            .collect_vec()
+                    });
+                });
+            instances?
+        };
+        debug_assert_eq!(instances.len(), num_instances);
+
+        Ok(Self {
+            svk: KzgSvk::<M>::new(params.get_g()[0]),
+            snark: SnarkWitness::new(
+                super_circuit_protocol,
+                super_circuit_instances,
+                super_circuit_proof,
+            ),
+            instances,
+        })
+    }
+
+    /// Returns accumulator indices in instance columns, which will be in
+    /// the last `4 * LIMBS` rows of instance column in `MainGate`.
+    pub fn accumulator_indices(&self) -> Vec<(usize, usize)> {
+        let offset = self.snark.protocol().num_instance.iter().sum::<usize>();
+        (offset..).map(|idx| (0, idx)).take(4 * LIMBS).collect()
+    }
+
+    /// Returns number of instance
+    pub fn num_instance(&self) -> Vec<usize> {
+        vec![self.snark.protocol().num_instance.iter().sum::<usize>() + 4 * LIMBS]
+    }
+
+    /// Returns instances
+    pub fn instances(&self) -> Vec<Vec<M::Scalar>> {
```

Comment:
Can you rename it to `fn instance` to match the other circuits?
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/a75c9aa3bda78575fcee1560cc6ff44f7aa9e22a/zkevm-circuits/src/super_circuit.rs#L293

Reply:
Thanks for catching this! Fixed in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1000/commits/2702faf90871d2cad169de930ce1f63d55e29462.